File: webapp/eXide/src/commands.js
/*
 *  eXist Open Source Native XML Database
 *  Copyright (C) 2011 The eXist Project
 *  http://exist-db.org
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *  $Id$
 */
eXide.namespace("eXide.edit.commands");

/**
 * Register editor commands to be called from keybindings.
 */
eXide.edit.commands = (function () {

	var canon = require("pilot/canon");
	var useragent = require("pilot/useragent");
	
	function bindKey(win, mac) {
	    return {
	        win: win,
	        mac: mac,
	        sender: "editor"
	    };
	}
	
	return {
		
		init: function (editor) {
		    canon.addCommand({
		    	name: "saveDocument",
		    	bindKey: bindKey("Ctrl-Shift-S", "Command-Shift-S"),
		    	exec: function (env, args, request) {
		    		eXide.app.saveDocument();
		    	}
		    });
		    canon.addCommand({
		    	name: "runQuery",
		    	bindKey: bindKey("Ctrl-Return", "Command-Return"),
		    	exec: function (env, args, request) {
		    		eXide.app.runQuery();
		    	}
		    });
		    canon.addCommand({
		    	name: "openDocument",
		    	bindKey: bindKey("Ctrl-Shift-O", "Command-Shift-O"),
		    	exec: function (env, args, request) {
		    		eXide.app.openDocument();
		    	}
		    });
		    canon.addCommand({
		    	name: "newDocument",
		    	bindKey: bindKey("Ctrl-Shift-N", "Command-Shift-N"),
		    	exec: function (env, args, request) {
		    		eXide.app.newDocument();
		    	}
		    });
		    canon.addCommand({
		    	name: "closeDocument",
		    	bindKey: bindKey("Ctrl-Shift-W", "Command-Shift-W"),
		    	exec: function (env, args, request) {
		    		eXide.app.closeDocument();
		    	}
		    });
		    canon.addCommand({
		    	name: "autocomplete",
		    	bindKey: bindKey("Ctrl-Space", "Ctrl-Space"),
		    	exec: function(env, args, request) {
		    		editor.autocomplete();
		    	}
		    });
		    canon.addCommand({
		    	name: "nextTab",
		    	bindKey: bindKey("Ctrl-Shift-PageDown", "Command-Shift-PageDown"),
		    	exec: function(env, args, request) {
		    		editor.nextTab();
		    	}
		    });
		    canon.addCommand({
		    	name: "previousTab",
		    	bindKey: bindKey("Ctrl-Shift-PageUp", "Command-Shift-PageUp"),
		    	exec: function(env, args, request) {
		    		editor.previousTab();
		    	}
		    });
		    canon.addCommand({
		    	name: "functionDoc",
		    	bindKey: bindKey("F1", "F1"),
		    	exec: function(env, args, request) {
		    		editor.exec("showFunctionDoc");
		    	}
		    });
		    canon.addCommand({
		    	name: "gotoDefinition",
		    	bindKey: bindKey("F3", "F3"),
		    	exec: function(env, args, request) {
		    		editor.exec("gotoDefinition");
		    	}
		    });
		    canon.addCommand({
		    	name: "indentOrParam",
		    	bindKey: bindKey("Tab", "Tab"),
		    	exec: function(env, args, request) {
		    		// if there's active template code in the document, tab will
		    		// cycle through the template's params. Otherwise, it calls indent.
		    		var doc = editor.getActiveDocument();
		    		if (!(doc.template && doc.template.nextParam())) {
		    			env.editor.indent();
		    		}
		    	}
		    });
		    canon.addCommand({
		    	name: "escape",
		    	bindKey: bindKey("Esc", "Esc"),
		    	exec: function(env, args, request) {
		    		var doc = editor.getActiveDocument();
		    		doc.template = null;
		    		env.editor.clearSelection();
		    	}
		    });
		    canon.addCommand({
		    	name: "dbManager",
		    	bindKey: bindKey("Ctrl-Shift-M", "Command-Shift-M"),
		    	exec: function (env, args, request) {
		    		eXide.app.manage();
		    	}
		    })
		},
		
		help: function (container) {
			$(container).find("table").each(function () {
				this.innerHTML = "";
				var names = canon.getCommandNames();
				for (var i = 0; i < names.length; i++) {
					var cmd = canon.getCommand(names[i]);
					var tr = document.createElement("tr");
					var td = document.createElement("td");
					td.appendChild(document.createTextNode(names[i]));
					tr.appendChild(td);
					td = document.createElement("td");
					if (useragent.isMac)
						td.appendChild(document.createTextNode(cmd.bindKey.mac));
					else
						td.appendChild(document.createTextNode(cmd.bindKey.win));
					tr.appendChild(td);
					this.appendChild(tr);
				}
			});
		}
	};
}());

File: src/org/exist/plugin/command/Commands.java
/*
 *  eXist Open Source Native XML Database
 *  Copyright (C) 2010 The eXist Project
 *  http://exist-db.org
 *  
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *  
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *  
 *  $Id$
 */
package org.exist.plugin.command;

import java.util.HashMap;
import java.util.Map;

import org.exist.xmldb.XmldbURI;

/**
 * Commands hub.
 * 
 * @author <a href="mailto:shabanovd@gmail.com">Dmitriy Shabanov</a>
 *
 */
public class Commands {
	
	private static Map<String, CommandResolver> commands = new HashMap<String, CommandResolver>();
	
	public static void plug(String command, CommandResolver resolver) {
		commands.put(command, resolver);
	}
	
	//XXX: plug-in activator
	static {
		plugin("org.exist.commands.info.InfoCommandResolver");
		plugin("org.exist.commands.threads.ThreadsCommandResolver");
		plugin("org.exist.commands.svn.SvnCommandResolver");
	}
	
	private static void plugin(String name) {
		try {
			Class.forName(name);
		} catch (final ClassNotFoundException e) {
		}
	}
	//*************************************************************************

	public static void command(XmldbURI collection, String[] params) throws CommandException {
		
		final CommandResolver runner = commands.get(params[0]);
		
		if (runner == null) {throw new CommandNotFoundException("Command '"+params[0]+"' not found.");}
		
    	final String[] commandData = new String[params.length-1];
    	System.arraycopy(params, 1, commandData, 0, params.length-1);
    	
    	runner.execute(collection, commandData);
	}

}
